README.md
# chef_client_updater

[![Build Status](https://travis-ci.org/chef-cookbooks/chef_client_updater.svg?branch=master)](https://travis-ci.org/chef-cookbooks/chef_client_updater) [![Cookbook Version](https://img.shields.io/cookbook/v/chef_client_updater.svg)](https://supermarket.chef.io/cookbooks/chef_client_updater)

This cookbook updates the chef-client

## Requirements

### Platforms

- All platforms with a chef-client package on downloads.chef.io

### Chef

- Chef 11.6.2+

## Usage

This cookbook provides both a custom resource and a default recipe. The default recipe simply uses the custom resource with a set of attributes. You can add chef_client_updater::default to your run list or use the custom resource in a wrapper cookbook.

### Init System Caveats

When Chef runs as a service under a system init daemon such as Sys-V or systemd each chef run forks off from the main chef-client process being managed by the init system. For a chef-client upgrade to occur the running chef-client as well as the parent process must be killed, and a new chef-client must start using the updated binaries. This cookbook handles killing the chef-client, but your init system must properly handle starting the service back up. For systemd and upstart this can be handled via configuration, and chef-client cookbook 8.1.1 or later handles this by default. This functionality is not available in sys-v (RHEL 6, Debian 7, and others), so you will need to employ a secondary process such as a monitoring system to start the chef-client service.

### Updating Windows Nodes

There are a couple of considerations on Windows that have to be dealt with. The Chef Client installer uses a custom component to speed up the installation. This component does not gracefully handle open file handles the way the MSI installer does. To work around this, the resource moves the currently installed Chef Client to a staging directory and that clears the way for the newer installer to run. At the end of that installation process though, that Chef Client run must exit or it will fail trying to find files that do not exist in their expected locations. The next run of the Chef Client will use the newly installed version.

#### Running Chef Client as a Scheduled Task

If you run as a scheduled task, then this will work smoothly. The path to the newly installed Chef Client will be the same and the scheduled task will launch it. Part of this resource's job on the next run is to make sure the staging directory with the older client is removed.

#### Running Chef Client As A Windows Service

If you run Chef Client as a service, things get a tiny bit more complicated. When the new installer runs, the service is removed. This isn't a big deal if you've got the chef-client cookbook set to configure the Windows service. If that is the case, we can register a scheduled task to run shortly after the `chef_client_updater` terminates the current chef run. An example recipe might look like:

```ruby
if Chef::VERSION < node['my_update_cookbook']['desired_version']
  run_chef_task_in_ten_minutes = Time.now + 600

  windows_task 'chef-client-upgrade' do
    cwd 'C:\\opscode\\chef\\bin'
    command 'chef-client'
    run_level :highest
    frequency :once
    start_time "#{run_chef_task_in_ten_minutes.strftime('%H:%M')}"
    action :create
  end
else
  windows_task 'chef-client-upgrade' do
    cwd 'C:\\opscode\\chef\\bin'
    command 'chef-client'
    run_level :highest
    frequency :once
    start_time "#{run_chef_task_in_ten_minutes.strftime('%H:%M')}"
    action :delete
  end
end

chef_client_updater 'Install latest Chef' do
  post_install_action 'kill'
end
```

### Upgrading from Chef 11

Moving from Chef 11 has a few challenges when we are dealing with public update sources. Chef 11 ships with a very old `cacert.pem`. To work through this, we need to get a more current `cacert.pem` file and point OpenSSL to it. Unfortunately, for this to work consistently on Windows, we'll need to reboot. Chef 11 does not have the reboot resource, so this isn't a graceful process. However, on the next Chef run after the reboot, things will be back on track and the upgrade will perform as on other platforms.

Below is an example of a recipe that can set up Chef 11 to work using public update sources.

```ruby
if platform_family?('windows') && (Chef::VERSION < '12')
  new_cert_file = File.join(ENV['USERPROFILE'], 'cacert.pem')

  remote_file new_cert_file do
    source 'https://curl.haxx.se/ca/cacert.pem'
    action :create
  end

  powershell_script 'restart' do
    code <<-EOH
    restart-computer -force
    EOH
    action :nothing
  end

  env 'SSL_CERT_FILE' do
    value new_cert_file
    notifies :run, 'powershell_script[restart]', :immediately
  end
end

chef_client_updater 'Install latest Chef' do
  post_install_action 'kill'
end
```

## Resources

### chef_client_updater

Installs the mixlib-install/mixlib-install gems and upgrades the chef-client.

#### properties

- `channel` - The chef channel you fetch the chef client from. `stable` contains all officially released chef-client builds where as `current` contains unreleased builds. Default: `stable`
- `prevent_downgrade` - Don't allow this cookbook to downgrade the chef-client version. Default: false
- `version` - The version of the chef-client to install. Default :latest
- `post_install_action` - After installing the chef-client what should we do. `exec` to exec the new client or `kill` to kill the client and rely on the init system to start up the new version. Default: `kill`
- `exec_command` - The chef-client command. default: $PROGRAM_NAME.split(' ').first
- `exec_args` - An array of arguments to exec the chef-client with. default: ARGV
- `download_url_override` - The direct URL for the chef-client package.
- `checksum` - The SHA-256 checksum of the chef-client package from the direct URL.
- `upgrade_delay` - The delay in seconds before the scheduled task to upgrade chef-client runs on windows. default: 30
- `product_name` - The name of the product to upgrade. This can be `chef` or `chefdk` default: chef

#### examples

```ruby
chef_client_updater 'Install latest'
```

```ruby
chef_client_updater 'Install latest Chef 13.x' do
  version '13'
end
```

```ruby
chef_client_updater 'Install 12.13.36 and kill' do
  version '12.13.36'
  post_install_action 'kill'
end
```

#### Test Kitchen Testing

In order to test this cookbook it will be necessary to change the `post_install_action` to `exec` from `kill`. While `kill` is better in most actual production use cases as it terminates the chef-client run along with cleaning up the parent process, the use of `kill` under test kitchen will fail the chef-client run and fail the test-kitchen run. The use of `exec` allows test-kitchen to complete and then re-runs the recipe to validate that the cookbook does not attempt to re-update the chef-client and will succeed with the new chef-client. This, however, means that it is not possible to exactly test the config which will be running in production. The best practice advice for this cookbook will be to ignore common best practices and not worry about that. If you change your production config to use `exec` in order to run what you test in test-kitchen, then you will find sharp edge cases where your production upgrades will hang and/or fail, which testing will not replicate. In order to test you should most likely test upgrades on your full-scale integration environment (not under test-kitchen) before rolling out to production and not use test-kitchen at all. If you think that there's a rule that you must test absolutely everything you run under test-kitchen, you should probably [read this](http://labs.ig.com/code-coverage-100-percent-tragedy) or [this](https://coderanger.net/overtesting/).

In order to test that your recipes work under the new chef-client codebase, you should simply test your cookbooks against the new version of chef-client that you wish to deploy in "isolation" from the upgrade process. If your recipes all work on the old client, and all work on the new client, and the upgrader works, then the sum of the parts should work as well (and again, if you really deeply care about the edge conditions where that might not work -- then test on real production-like images and not with test-kitchen).

#### Use of 'exec' in production

This is highly discouraged since the exec will not clean up the supervising process. You're very likely to see it upgrade successfully and then see the old chef-client process continue to run and fork off copies of the old chef-client to run again. Or for the upgrade process to hang, or for other issues to occur causing failed upgrades.

You can use 'exec' in production if you are running from cron or some other process manager and firing off single-shot `--no-fork` chef-client processes without using the `--interval` option. This will have the advantage that the new chef-client kicks off immediately after the upgrade giving fast feedback on any failures under the new chef-client. The utility of this approach is most likely is not enough to justify the hassle.

## A note about purpose

While this cookbook supports running on Chef versions back to 11/12, the supported behavior of the cookbook is to upgrade those versions to 13/14 or newer.  It is not intended that users would maintain old Chef-11/12 versions with this
cookbook.  The latest released version of Chef 12 (12.22.1 or later) is still be supported as a target.  Older versions of omnibus chef will have their embedded rubygems force upgraded by this cookbook to avoid having to regression test against
5+ years of rubygems bugs and establish a stable basis for the cookbook to use.

## Maintainers

This cookbook is maintained by Chef's Community Cookbook Engineering team. Our goal is to improve cookbook quality and to aid the community in contributing to cookbooks. To learn more about our team, process, and design goals see our [team documentation](https://github.com/chef-cookbooks/community_cookbook_documentation/blob/master/COOKBOOK_TEAM.MD). To learn more about contributing to cookbooks like this see our [contributing documentation](https://github.com/chef-cookbooks/community_cookbook_documentation/blob/master/CONTRIBUTING.MD), or if you have general questions about this cookbook come chat with us in #cookbok-engineering on the [Chef Community Slack](http://community-slack.chef.io/)

## License

```text
Copyright:: 2016-2018, Chef Software, Inc

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
```


CHANGELOG.md
# chef_client_updater Cookbook CHANGELOG

This file is used to list changes made in each version of the chef_client_updater cookbook.

## 3.4.4 (2018-08-31)

Backport to fork
- Only run the cron job on *nix 5 minutes from now

## 3.4.3 (2018-08-20)

- Allow this to work in an air gapped environment that doesn't have a gem server

## 3.4.2 (2018-08-15)

- Fix to retrieve current version for angrychef nodes

## 3.4.1 (2018-08-02)

- Allow for a configurable package install timeout. (#130)

## 3.4.0 (2018-07-23)

- Prevent failures on Chef-DK
- rubygems_url: Allow the rubygems source to be specified
- Require mixlib-install 3.11+ for Amazon Linux 2.0 support
- Attempt to move the chef install directory if it still exists before giving up to workaround failures on Windows 2008

## 3.3.5 (2018-06-20)

- Do not attempt EventLog restart on Windows 7 or Windows Server 2008 editions.

## 3.3.4 (2018-05-30)

- Improve install success rate on Windows by destroying existing Chef handles prior to upgrade.

## 3.3.3 (2018-04-26)

- Better support AIX by also restarting the chef-client service via cron on AIX

## 3.3.2 (2018-04-11)

- Improve how we handle updating rubygems:
  - Update RubyGems without shelling out
  - Don't upgrade to the very latest version unless we have to (due to old current rubygems)

## 3.3.1 (2018-04-11)

- Stop push jobs before upgrading Chef on Windows and then start it back up. This prevents failures to update.

## 3.3.0 (2018-04-10)

- Post action support on Windows
- Prevent failure when the user doesn't use the chef-client cookbook

## 3.2.9 (2018-04-05)

- Don't run cron on Windows

## 3.2.8 (2018-04-04)

- Restart chef-client with cron under sysvinit to ensure that chef-client properly restarts

## 3.2.7 (2018-03-29)

- Update log message to KILL from TERM. We updated the behavior but missed the logging

## 3.2.6 (2018-03-16)

- Added additional logic to decide if a 'kill' or 'exec' should be done post upgrade. If the chef-client isn't running with the supervsior process then we will no longer try to use kill.

## 3.2.5 (2018-02-28)

- Use KILL instead of TERM on Windows since TERM isn't always actually killing the process
- Updated the upgrade_delay value on Windows to be 60 seconds since anything less causes issues

## 3.2.4 (2018-02-05)

- Fix warning '/ST is earlier than current time' on Windows

## 3.2.3 (2018-01-25)

- Make product_name attribute into a real property on the resource

## 3.2.2 (2018-01-25)

- Make upgrade_delay a true resource property

## 3.2.1 (2018-01-25)

- Added new attribute 'upgrade_delay' that defines delay in seconds before upgrading Chef client.

## 3.2.0 (2018-01-22)

- Require mixlib-install 3.9 which includes initial support for proxies and support for Amazon Linux 2.0
- Add additional debug logging for the installation
- If the user provides an X.Y.Z format version don't contact Chef Inc's servers to validate the version
- error out chef run if shell update fails

## 3.1.3 (2017-12-20)

- Support custom paths to chef-client

## 3.1.2 (2017-11-22)

- Windows: Support for downgrading the Chef client

## 3.1.1 (2017-11-14)

- Windows: Bypass Powershell Execution Policy for upgrade scheduled task

## 3.1.0 (2017-11-01)

- Raise if download_url_override is set but checksum isn't
- Require a mixin-install that supports proxies
- Improve how we perform the cleanup of the previous install directory on Windows
- Remove a hardcoded path to the chef-client on Windows
- Improve how we perform the Windows upgrade by using a scheduled task to avoid failures during the upgrade

## 3.0.4 (2017-08-17)

- Fix gem install to actually install mixlib-install 3.3.4
- Fix :latest resulting in chef installing on every run

## 3.0.3 (2017-08-10)

- Add accurate logging for the rubygems upgrade to reflect that we're actually upgrading to the latest release.
- Require mixlib-install 3.3.4 to prevent failures on Windows nodes due to architecture detection failing within mixlib-install
- Add debug logging for the desired version logic
- Improve logging of the version we're upgrading to in situations where the user provides either :latest or a partial version like '12'. Show the version we're upgrading to instead of what the user passed

## 3.0.2 (2017-08-08)

- Improve logging to actually log when the upgrade occurs before we kill or exec the current chef-client run

## 3.0.1 (2017-07-14)

- adding check for gem.cmd on chef-client 13 windows

## 3.0.0 (2017-07-14)

### Breaking Changes

- The default post install action for the resource has been switched from exec to kill. We believe that kill is the most likely default that users would expect as this allows a chef-client daemon running under a modern init system to cleanly upgrade. We highly recommend you check the readme and understand the exec and kill functions to determine which makes the most sense for how you run chef-client and how you execute the upgrade.
- The prevent downgrade attribute for the default recipe has been changes from true to false as this is the default for the resource.

### Other Changes

- If chef-client is forked and the user specifies an 'exec' post install action we will now warn and then kill as exec will not worked in a forked run
- Updated the windows task example in the readme to properly set the start time
- Updated the minimum version of mixlib-install to download 3.3.1 which includes many fixes for windows clientsA
- The resource now works around situations where mixlib-install may return an Array of versions

## 2.0.3 (2017-06-27)

- Fix #31 detect centos platform correctly

## 2.0.2 (2017-06-22)

- Fix air-gapped installation regression introduced by support for partial versions

## 2.0.1 (2017-06-16)


providers/default.rb
#
# Cookbook:: chef_client_updater
# Resource:: updater
#
# Copyright:: 2016-2018, Will Jordan
# Copyright:: 2016-2018, Chef Software Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

# NOTE: this cookbook uses Chef-11 backwards compatible syntax to support
# upgrades from Chef 11.x and this pattern should not be copied for any modern
# cookbook.  This is a poor example cookbook of how to write Chef.

use_inline_resources # ~FC113

provides :chef_client_updater if respond_to?(:provides)

def load_mixlib_install
  gem 'mixlib-install', '~> 3.11'
  require 'mixlib/install'
rescue LoadError
  Chef::Log.info('mixlib-install gem not found. Installing now')
  chef_gem 'mixlib-install' do
    version '~> 3.11'
    compile_time true if respond_to?(:compile_time)
    if new_resource.rubygems_url
      clear_sources true if respond_to?(:clear_sources)
      options "--source #{new_resource.rubygems_url}" if respond_to?(:options)
    end
  end

  require 'mixlib/install'
end

def load_mixlib_versioning
  require 'mixlib/versioning'
rescue LoadError
  Chef::Log.info('mixlib-versioning gem not found. Installing now')
  chef_gem 'mixlib-versioning' do
    compile_time true if respond_to?(:compile_time)
    if new_resource.rubygems_url
      clear_sources true if respond_to?(:clear_sources)
      options "--source #{new_resource.rubygems_url}" if respond_to?(:options)
    end
  end

  require 'mixlib/versioning'
end

def update_rubygems
  compatible_rubygems_versions = '>= 2.6.11'
  target_version = '2.6.11'

  rubygems_version = Gem::Version.new(Gem::VERSION)
  Chef::Log.debug("Found gem version #{rubygems_version}. Desired version is at least #{target_version}")
  return if Gem::Requirement.new(compatible_rubygems_versions).satisfied_by?(rubygems_version)

  # only rubygems >= 1.5.2 supports pinning, and we might be coming from older versions
  pin_rubygems_range = '>= 1.5.2'
  pin = Gem::Requirement.new(pin_rubygems_range).satisfied_by?(rubygems_version)

  converge_by "upgrade rubygems #{rubygems_version} to #{pin ? target_version : 'latest'}" do
    if new_resource.rubygems_url
      gem_bin = "#{Gem.bindir}/gem"
      if !::File.exist?(gem_bin) && windows?
        gem_bin = "#{Gem.bindir}/gem.cmd" # on Chef Client 13+ the rubygem executable is gem.cmd, not gem
      end
      raise 'cannot find omnibus install' unless ::File.exist?(gem_bin)
      source = "--clear-sources --source #{new_resource.rubygems_url}"
      shell_out!("#{gem_bin} update --system --no-rdoc --no-ri #{source}")
    else
      require 'rubygems/commands/update_command'
      args = ['--no-rdoc', '--no-ri', '--system' ]
      args.push(target_version) if pin
      Gem::Commands::UpdateCommand.new.invoke(*args)
    end
  end
end

def load_prerequisites!
  # This doesn't work in airgapped environments without a gem server
  # update_rubygems
  load_mixlib_install
  load_mixlib_versioning
end

def mixlib_install
  load_mixlib_install
  detected_platform = Mixlib::Install.detect_platform
  Chef::Log.debug("Platform detected as #{detected_platform} by mixlib_install")
  options = {
    product_name: new_resource.product_name,
    platform_version_compatibility_mode: true,
    platform: detected_platform[:platform],
    platform_version: detected_platform[:platform_version],
    architecture: detected_platform[:architecture],
    channel: new_resource.channel.to_sym,
    product_version: new_resource.version == 'latest' ? :latest : new_resource.version,
  }
  if new_resource.download_url_override
    raise('Using download_url_override in the chef_client_updater resource requires also setting checksum property!') unless new_resource.checksum
    Chef::Log.debug("Passing download_url_override of #{new_resource.download_url_override} and checksum of #{new_resource.checksum} to mixlib_install")
    options[:install_command_options] = { download_url_override: new_resource.download_url_override, checksum: new_resource.checksum }
  end
  Chef::Log.debug("Passing options to mixlib-install: #{options}")
  Mixlib::Install.new(options)
end

# why would we use this when mixlib-install has a current_version method?
# well mixlib-version parses the manifest JSON, which might not be there.
# ohai handles this better IMO
# @return String
def current_version
  case new_resource.product_name
  when 'chef', 'angrychef'
    node['chef_packages']['chef']['version']
  when 'chefdk'
    Mixlib::ShellOut.new('chef -v').run_command.stdout.match(/^Chef Development Kit Version.*/).to_s.split(': ').last
  end
end

# the version we WANT TO INSTALL. If the user specifies a version in X.Y.X format
# we use that without looking it up. If :latest or a non-X.Y.Z format version we
# look it up with mixlib-install to determine the latest version matching the request
# @return Mixlib::Versioning::Format::PartialSemVer
def desired_version
  if new_resource.version.to_sym == :latest # we need to find what :latest really means
    version = Mixlib::Versioning.parse(mixlib_install.available_versions.last)
    Chef::Log.debug("User specified version of :latest. Looking up using mixlib-install. Value maps to #{version}.")
  elsif new_resource.download_url_override # probably in an air-gapped environment.
    version = Mixlib::Versioning.parse(new_resource.version)
    Chef::Log.debug("download_url_override specified. Using specified version of #{version}")
  elsif new_resource.version.split('.').count == 3 # X.Y.Z version format given
    Chef::Log.debug("User specified version of #{new_resource.version}. No need check this against Chef servers.")
    version = Mixlib::Versioning.parse(new_resource.version)
  else # lookup their shortened version to find the X.Y.Z version
    version = Mixlib::Versioning.parse(Array(mixlib_install.artifact_info).first.version)
    Chef::Log.debug("User specified version of #{new_resource.version}. Looking up using mixlib-install as this is not X.Y.Z format. Value maps to #{version}.")
  end
  version
end

# why wouldn't we use the built in update_available? method in mixlib-install?
# well that would use current_version from mixlib-install and it has no
# concept of preventing downgrades
def update_necessary?
  load_mixlib_versioning
  cur_version = Mixlib::Versioning.parse(current_version)
  des_version = desired_version

  Chef::Log.debug("The current chef-client version is #{cur_version} and the desired version is #{des_version}")
  necessary = new_resource.prevent_downgrade ? (des_version > cur_version) : (des_version != cur_version)
  Chef::Log.debug("A chef-client upgrade #{necessary == true ? 'is' : "isn't"} necessary")
  necessary
end

def eval_post_install_action
  return unless new_resource.post_install_action == 'exec'

  if Chef::Config[:interval] || Chef::Config[:client_fork]
    Chef::Log.warn 'post_install_action "exec" not supported for chef-client running forked -- changing to "kill".'
    new_resource.post_install_action = 'kill'
  end

  return unless windows?

  Chef::Log.warn 'forcing "exec" to "kill" on windows.'
  new_resource.post_install_action = 'kill'
end

def run_post_install_action
  # make sure the passed action will actually work
  eval_post_install_action

  case new_resource.post_install_action
  when 'exec'
    if Chef::Config[:local_mode]
      Chef::Log.info 'Shutting down local-mode server.'
      if Chef::Application.respond_to?(:destroy_server_connectivity)
        Chef::Application.destroy_server_connectivity
      elsif defined?(Chef::LocalMode) && Chef::LocalMode.respond_to?(:destroy_server_connectivity)
        Chef::LocalMode.destroy_server_connectivity
      end
    end
    Chef::Log.warn 'Replacing chef-client process with upgraded version and re-running.'
    Kernel.exec(new_resource.exec_command, *new_resource.exec_args)
  when 'kill'
    if Chef::Config[:client_fork] && Process.ppid != 1 && !windows?
      Chef::Log.warn 'Chef client is running forked with a supervisor. Sending KILL to parent process!'
      Process.kill('KILL', Process.ppid)
    end
    Chef::Log.warn 'New chef-client installed and exit is allowed. Forcing chef exit!'
    exit(213)
  else
    raise "Unexpected post_install_action behavior: #{new_resource.post_install_action}"
  end
end

def chef_install_dir
  node['chef_client_updater']['chef_install_path'] || (windows? ? 'c:/opscode/chef' : '/opt/chef')
end

def chef_backup_dir
  "#{chef_install_dir}.upgrade"
end

def chef_broken_dir
  "#{chef_install_dir}.broken"
end

def chef_upgrade_log
  "#{chef_install_dir}_upgrade.log"
end

# cleanup cruft from *prior* runs
def cleanup
  if ::File.exist?(chef_backup_dir)
    converge_by("remove #{chef_backup_dir} from previous chef-client run") do
      FileUtils.rm_rf chef_backup_dir
    end
  end
  if ::File.exist?(chef_upgrade_log)
    converge_by("remove #{chef_upgrade_log} from previous chef-client run") do
      FileUtils.rm_rf chef_upgrade_log
    end
  end
  # When running under init this cron job is created after an update
  cron 'chef_client_updater' do
    action :delete
  end unless platform_family?('windows')
end

def windows?
  platform_family?('windows')
end

def copy_opt_chef(src, dest)
  FileUtils.mkdir dest
  FileUtils.cp_r "#{src}/.", dest
rescue
  nil
end

# windows does not like having running open files nuked behind it so we have to move the old file
# out of the way.  on both platforms we must clean up the old install to not leave behind any old
# gem files.
#
def move_opt_chef(src, dest)
  converge_by("move all files under #{src} to #{dest}") do
    FileUtils.rm_rf dest
    raise "rm_rf of #{dest} failed" if ::File.exist?(dest) # detect mountpoints that were not deleted
    FileUtils.mv src, dest
  end
rescue => e
  # this handles mountpoints
  converge_by("caught #{e}, falling back to copying and removing from #{src} to #{dest}") do
    begin
      FileUtils.rm_rf dest
    rescue
      nil
    end # mountpoints can throw EBUSY
    begin
      FileUtils.mkdir dest
    rescue
      nil
    end # mountpoints can throw EBUSY
    FileUtils.cp_r Dir.glob("#{src}/*"), dest
    FileUtils.rm_rf Dir.glob("#{src}/*")
  end
end

def upgrade_start_time
  shifted_time = Time.now + new_resource.upgrade_delay
  shifted_time.strftime('%H:%M')
end

def prepare_windows
  copy_opt_chef(chef_install_dir, chef_backup_dir)
  Kernel.spawn("c:/windows/system32/schtasks.exe /F /RU SYSTEM /create /sc once /ST \"#{upgrade_start_time}\" /tn Chef_upgrade /tr \"powershell.exe -ExecutionPolicy Bypass c:/opscode/chef_upgrade.ps1 > #{chef_upgrade_log}\"")
  FileUtils.rm_rf "#{chef_install_dir}/bin/chef-client.bat"
end

def uninstall_ps_code
  uninstall_ps_code = <<-EOH
    function guid_from_regvalue($value) {
      $order = 7,6,5,4,3,2,1,0,11,10,9,8,15,14,13,12,17,16,19,18,21,20,23,22,25,24,27,26,29,28,31,30
      $dash_pos = 8,13,18,23

      $guid = ""
      $order | % {
        $letter = $value.Substring($_,1)
        $guid = "$guid$letter"
        if ($dash_pos -contains $guid.length) {$guid = "$guid-"}
      }
      return $guid
    }

    function installed_remove() {
      $installed_product = (get-item HKLM:\\SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\Installer\\UpgradeCodes\\C58A706DAFDB80F438DEE2BCD4DCB65C).Property
      $msi_guid = guid_from_regvalue $installed_product

      Start-Process msiexec.exe -Wait -ArgumentList "/x {$msi_guid} /q"
    }

    installed_remove
  EOH

  uninstall_ps_code
end

def open_handle_functions
  <<-EOH
  Function Get-OpenHandle {
    param(
          [Parameter(ValueFromPipelineByPropertyName=$true)]
          $Search
    )
    $handleOutput = &#{Chef::Config[:file_cache_path]}/handle.exe -accepteula -nobanner -a -u $Search
    $handleOutput | foreach {
      if ($_ -match '^(?<program>\\S*)\\s*pid: (?<pid>\\d*)\\s*type: (?<type>\\S*)\\s*(?<user>\\S*)\\s*(?<handle>\\S*):\\s*(?<file>(\\\\\\\\)|([a-z]:).*)') {
        $matches | select @{n="User";e={$_.user}},@{n="Path";e={$_.file}},@{n="Handle";e={$_.handle}},@{n="Type";e={$_.type}},@{n="HandlePid";e={$_.pid}},@{n="Program";e={$_.program}}
      }
    }
  }

  Function Destroy-Handle {
    param(
          [Parameter(Mandatory=$true,ValueFromPipeline=$true)]
          $Handle,
          [Parameter(Mandatory=$true,ValueFromPipeline=$true)]
          $HandlePid
    )

    $handleOutput = &#{Chef::Config[:file_cache_path]}/handle.exe -accepteula -nobanner -c $Handle -p $HandlePid -y
    '      Destroyed handle {0} from pid {1}' -f $Handle, $HandlePid | echo
  }

  Function Destroy-OpenChefHandles {
    echo '[*] Destroying open Chef handles.'
    Get-OpenHandle -Search opscode | foreach {
      '  [+] Destroying handle that {0} (pid: {1}) has on {2}' -f $_.Program, $_.HandlePid, $_.Path | echo
      Destroy-Handle -Handle $_.Handle -HandlePid $_.HandlePid
    }
    echo '[*] Completed destroying open Chef handles.'
  }
  EOH
end

def execute_install_script(install_script)
  if windows?
    cur_version = Mixlib::Versioning.parse(current_version)
    uninstall_first = if desired_version < cur_version
                        uninstall_ps_code
                      else
                        ''
                      end

    post_action = if new_resource.post_install_action == 'exec'
                    new_resource.exec_command
                  else
                    ''
                  end

    remote_file "#{Chef::Config[:file_cache_path]}/handle.zip" do
      source new_resource.handle_zip_download_url
      not_if { ::File.file?("#{Chef::Config[:file_cache_path]}/handle.exe") }
    end.run_action(:create)

    powershell_script 'name' do
      code <<-EOH
        $command = {
          Get-Service chef-client -ErrorAction SilentlyContinue | stop-service
          Get-Service push-jobs-client -ErrorAction SilentlyContinue | stop-service

          if ((Get-WmiObject Win32_Process -Filter "name = 'ruby.exe'" | Select-Object CommandLine | select-string 'opscode').count -gt 0) {
            Write-Output "Chef cannot be upgraded while in use. Exiting..."
            exit 8
          }

          if (!(Test-Path "#{Chef::Config[:file_cache_path]}/handle.exe")) {
            Add-Type -AssemblyName System.IO.Compression.FileSystem
            [System.IO.Compression.ZipFile]::ExtractToDirectory("#{Chef::Config[:file_cache_path]}/handle.zip", "#{Chef::Config[:file_cache_path]}")
          }

          #{open_handle_functions}

          if (Test-Path "#{Chef::Config[:file_cache_path]}/handle.exe") {
            Destroy-OpenChefHandles
          }

          $windows_kernel_version = (Get-WmiObject -class Win32_OperatingSystem).Version
          if (-Not ($windows_kernel_version.Contains('6.0') -or $windows_kernel_version.Contains('6.1'))) {
            Get-Service EventLog | Restart-Service -Force
          }

          Remove-Item "#{chef_install_dir}" -Recurse -Force

          if (test-path "#{chef_install_dir}") {
            Write-Output "Removing #{chef_install_dir} did not completely succeed."
            Write-Output "It is likely now in a bad state, not even usable as a backup."
            Write-Output "Attempting to move #{chef_install_dir} to #{chef_broken_dir}"
            Move-Item "#{chef_install_dir}" "#{chef_broken_dir}"
          }

          if (test-path "#{chef_install_dir}") {
            Write-Output "#{chef_install_dir} still exists, upgrade will be aborted. Exiting (3)..."
            exit 3
          }

          #{uninstall_first}
          #{install_script}

          Remove-Item "c:/opscode/chef_upgrade.ps1"
          c:/windows/system32/schtasks.exe /delete /f /tn Chef_upgrade

          #{post_action}

          Get-Service push-jobs-client -ErrorAction SilentlyContinue | start-service
        }

        $http_proxy = $env:http_proxy
        $set_proxy = "`$env:http_proxy=`'$http_proxy`'"

        Set-Content -Path c:/opscode/chef_upgrade.ps1 -Value $set_proxy
        Add-Content c:/opscode/chef_upgrade.ps1 "`n$command"

      EOH
      action :nothing
    end.run_action(:run)
  else
    upgrade_command = Mixlib::ShellOut.new(install_script, timeout: new_resource.install_timeout)
    upgrade_command.run_command
    if upgrade_command.exitstatus != 0
      raise "Error updating chef-client. exit code: #{upgrade_command.exitstatus}.\nSTDERR: #{upgrade_command.stderr}\nSTDOUT: #{upgrade_command.stdout}"
    end
  end
end

action :update do
  begin
    load_prerequisites!

    if update_necessary?
      converge_by "upgrade chef-client #{current_version} to #{desired_version}" do
        # we have to get the script from mibxlib-install..
        install_script = mixlib_install.install_command
        # ...before we blow mixlib-install away
        platform_family?('windows') ? prepare_windows : move_opt_chef(chef_install_dir, chef_backup_dir)

        execute_install_script(install_script)
      end
      converge_by 'take post install action' do
        run_post_install_action
      end
    else
      cleanup
    end
  rescue SystemExit
    # sysvinit won't restart after we exit, potentially use cron to do so
    # either trust the chef-client cookbook's init scripts or the users choice
    if (node['chef_client'] && node['chef_client']['init_style'] == 'init') || node['chef_client_updater']['restart_chef_via_cron']
      Chef::Log.warn 'Chef Client was upgraded, scheduling chef-client start via cron in 5 minutes'
      cron_time = Time.now + 300
      start_cmd = case node['platform_family']
                  when 'aix'
                    '/usr/bin/startsrc -s chef > /dev/console 2>&1'
                  else
                    '/etc/init.d/chef-client start'
                  end

      r = cron 'chef_client_updater' do
        hour cron_time.hour
        minute cron_time.min
        command start_cmd
      end

      r.run_action(:create)
    end

    raise
  rescue Exception => e # rubocop:disable Lint/RescueException
    if ::File.exist?(chef_backup_dir)
      Chef::Log.warn "CHEF UPGRADE ABORTED due to #{e}: rolling back to #{chef_backup_dir} copy"
      move_opt_chef(chef_backup_dir, chef_install_dir) unless platform_family?('windows')
    else
      Chef::Log.warn "NO #{chef_backup_dir} DIR TO ROLL BACK TO!"
    end
    raise
  end
end
